### PR TITLE
Properly handle lifetimes on erased trait generation

### PR DESCRIPTION
**Pull request by spastorino**
r? @tmandry

### CODE REVIEW

File: dynosaur/tests/pass/basic.stdout [hunk 1]
```diff
@@ -7,13 +7,17 @@ use dynosaur::dynosaur;
...
 trait ErasedMyTrait {
     type Item;
-    fn foo(&self)
-    -> ::core::pin::Pin<Box<dyn ::core::future::Future<Output = Self::Item>>>;
+    fn foo<'async_trait>()
+    ->
+        ::core::pin::Pin<Box<dyn ::core::future::Future<Output = Self::Item> +
+        'async_trait>>
+    where
+    Self: 'async_trait;
```
Comment (tmandry) [inline dynosaur/tests/pass/basic.stdout:20]: Unsure if we should be adding this lifetime on a method that doesn't take self as a parameter.

File: dynosaur_derive/src/lib.rs [hunk 2]
```diff
@@ -56,45 +65,166 @@ pub fn dynosaur(
...
-                    })
-                } else {
-                    item.clone()
-                }
-            })
-            .collect(),
-        ..item.clone()
-    };
+                    ) = item
+                    {
```
Comment (tmandry) [inline dynosaur_derive/src/lib.rs:86]: Maybe split this into a separate function?

File: dynosaur_derive/src/lib.rs [hunk 3]
```diff
@@ -56,45 +65,166 @@ pub fn dynosaur(
...
+                            .predicates
+                            .push(parse_quote! {
+                                Self: 'async_trait
+                            });
+
+                        for (i, arg) in sig.inputs.iter_mut().enumerate() {
+                            match arg {
+                                FnArg::Receiver(receiver) => {
+                                    if receiver.reference.is_none() {
+                                        receiver.mutability = None;
```
Comment (tmandry) [inline dynosaur_derive/src/lib.rs:166]: What's going on here?

File: dynosaur_derive/src/lib.rs [hunk 4]
```diff
@@ -56,45 +65,166 @@ pub fn dynosaur(
...
+                                    }
+                                }
+                                FnArg::Typed(arg) => {
+                                    if match *arg.ty {
+                                        Type::Reference(_) => false,
+                                        _ => true,
+                                    } {
+                                        if let Pat::Ident(pat) = &mut *arg.pat {
+                                            pat.by_ref = None;
+                                            pat.mutability = None;
```
Comment (tmandry) [inline dynosaur_derive/src/lib.rs:176]: And here?

File: dynosaur_derive/src/lib.rs [hunk 5]
```diff
@@ -56,45 +65,166 @@ pub fn dynosaur(
...
+                                        }
+                                    }
+                                    AddLifetimeToImplTrait.visit_type_mut(&mut arg.ty);
+                                }
+                            }
+                        }
+
+                        sig.output = parse_quote! {
+                            #ret_arrow ::core::pin::Pin<Box<
+                            dyn ::core::future::Future<Output = #ret> + 'async_trait>>
```
Comment (tmandry) [inline dynosaur_derive/src/lib.rs:190]: Name suggestion: `'dynosaur` :)

File: dynosaur/tests/pass/basic.stdout (hunk 1 above)
Comment (spastorino) [reply dynosaur/tests/pass/basic.stdout]: There's Self in the `Output =` part. Anyway, you're right in the sense that I'm not correctly checking if the function has self or not. Addressed that and added a new test.

File: dynosaur_derive/src/lib.rs (hunk 3 above)
Comment (spastorino) [reply dynosaur_derive/src/lib.rs]: Removed as this is async_trait's code that I'm not 100% why is there. If that is really needed we can add it later.

File: dynosaur_derive/src/lib.rs (hunk 4 above)
Comment (spastorino) [reply dynosaur_derive/src/lib.rs]: Same here.

File: dynosaur/tests/pass/basic-with-self.stdout [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    -> Self::Item;
+}
+trait ErasedMyTrait {
+    type Item;
+    fn foo<'dynosaur>()
+    ->
+        ::core::pin::Pin<Box<dyn ::core::future::Future<Output = Self::Item> +
+        'dynosaur>>
+    where
+    Self: 'dynosaur;
```
Comment (tmandry) [inline dynosaur/tests/pass/basic-with-self.stdout:20]: It looks like you tried to remove this, but it's still here? Maybe it's seeing `Self::Item` in the signature. But the only thing the where clause would be good for is allowing us to capture arguments in the returned future, and there are no arguments.

**Review by tmandry — APPROVED** [review]

**Pull request merged**